### ISSUE TITLE
PPO takes very long

### ISSUE THREAD

**Issue by mandyyyyii** [OpenRLHF/OpenRLHF#420]
I have 3 questions.

1. PPO takes about 30~ hr and it only reaches to 40 steps. I have pasted my command. Is this normal? How long does it take to run PPO?
set -x export PATH=$HOME/.local/bin/:$PATH

ray job submit --address="http://127.0.0.1:8265" \
    --runtime-env-json='{"working_dir": "/openrlhf", "pip": "/openrlhf/requirements.txt"]}' \
    -- python3 examples/train_ppo_ray.py \
    --ref_num_nodes 1 \
    --ref_num_gpus_per_node 1 \
    --reward_num_nodes 1 \
    --reward_num_gpus_per_node 1 \
    --critic_num_nodes 1 \
    --critic_num_gpus_per_node 1 \
    --actor_num_nodes 1 \
    --actor_num_gpus_per_node 4 \
    --pretrain OpenLLMAI/Llama-2-7b-sft-model-ocra-500k \
    --reward_pretrain OpenLLMAI/Llama-2-7b-rm-anthropic_hh-lmsys-oasst-webgpt \
    --save_path /openrlhf/examples/test_scripts/ckpt/7b_llama \
    --micro_train_batch_size 2 \
    --train_batch_size 128 \
    --micro_rollout_batch_size 8 \
    --rollout_batch_size 1024 \
    --max_epochs 1 \
    --prompt_max_len 1024 \
    --generate_max_len 1024 \
    --zero_stage 2 \
    --bf16 \
    --actor_learning_rate 5e-7 \
    --critic_learning_rate 9e-6 \
    --init_kl_coef 0.01 \
    --prompt_data Open-Orca/OpenOrca,Dahoas/full-hh-rlhf,tasksource/oasst1_pairwise_rlhf_reward \
    --prompt_data_probs 0.4,0.5,0.1 \
    --max_samples 80000 \
    --normalize_reward \
    --adam_offload \
    --flash_attn \
    --gradient_checkpointing \
    --use_wandb {wandb_token}

2. The implementation for reward model seems to initialize a linear layer. Since we don't train reward model, isn't it causing some randomness? Why not just use the reward model?

3. The critic model is initialized from reward model. Does it make much difference to initialize from policy model or reward model?

**Comment by hijkzzz** [comment]
1. Could you please provide your GPU information? If you use vLLM acceleration, you can achieve double the speed.
2. We didn't randomly initialize the final layer in PPO; using RM to initialize the critic leads to faster convergence.

**Comment by mandyyyyii** [comment]
Yes I use vllm and I make the default of vllm as 1. I use 8 A100 40gb.

**Comment by mandyyyyii** [comment]
Why not just use the default reward model but change final layer? (https://github.com/OpenRLHF/OpenRLHF/blob/e1205adaf163e28ed58400ed84618bc01c3c4c37/openrlhf/models/model.py#L187)

**Comment by hijkzzz** [comment]
> Why not just use the default reward model but change final layer? (
> 
> https://github.com/OpenRLHF/OpenRLHF/blob/e1205adaf163e28ed58400ed84618bc01c3c4c37/openrlhf/models/model.py#L187
> 
> )

self.value_head is also initialized from the rm checkpoint.

**Comment by mandyyyyii** [comment]
ok, could I ask how much time it would take to run the ppo normally? It has been 48 hours and only gone to 60 steps.

**Comment by hijkzzz** [comment]
> ok, could I ask how much time it would take to run the ppo normally? It has been 48 hours and only gone to 60 steps.

you should increase the GPUs for vLLM. See https://github.com/OpenRLHF/OpenRLHF/blob/main/examples/scripts/train_ppo_llama_ray.sh

**Comment by mandyyyyii** [comment]
How much time does it take to train ppo usually? Would increasing number of GPUs for vLLM speed up? Even increasing number of gpus for vllm, I am not sure if that would help that much. 

For that file, it totally uses 10 gpu. If we only have 8 gpus, how should it be distributed?

**Comment by hijkzzz** [comment]
> How much time does it take to train ppo usually? Would increasing number of GPUs for vLLM speed up? Even increasing number of gpus for vllm, I am not sure if that would help that much.
> 
> For that file, it totally uses 10 gpu. If we only have 8 gpus, how should it be distributed?

We have enabled node consolidation:
```
   --colocate_critic_reward \
   --colocate_actor_ref \
```

This will speed up a lot and be able to train within one day.

**Comment by mandyyyyii** [comment]
Sorry for my unclear initial question. To clarify:
I'm currently using vLLM with 1 GPU. I'm wondering if increasing to 2 GPUs for vLLM would provide a significant performance improvement. My current setup allocates GPUs as follows:

4 GPUs for the actor
1 GPU each for reference, critic, reward, and vLLM

This configuration is resulting in long processing times. I'm also curious if implementing colocation might help improve overall performance. Thank you.

**Comment by hijkzzz** [comment]
Please use
4 for vLLM, 2 for actor (shared with ref + ref_offload), 2 for critic (shared with rm + rm_offload)
and enable adam_offload
